### PR TITLE
Handle AvoidMap recursions

### PR DESCRIPTION
**Pull request by odersky**
Issue a diagnostic instead of crashing when there is a stack overflow

Avoids the crash in the minimized version of #15365

### CODE REVIEW

**Comment by odersky** [comment]
We still need to get to the bottom of why we get the cyclic lower bounds that lead to the problem.

**Comment by smarter** [comment]
> We still need to get to the bottom of why we get the cyclic lower bounds that lead to the problem.

I have a simpler example where the constraint solver can create cycles like this: https://github.com/lampepfl/dotty/issues/15280, in this issue I proposed to detect the cycle at instantiation and emit an error then, but this isn't enough for operations like `avoid` that traverse the bounds of uninstantatiated type variables. So unless by some miracle we can come up with a cycle-free constraint solver, I guess we'd have to guard recursion in `avoid` and other operations by keeping track of seen type variables.

File: tests/neg-custom-args/allow-deep-subtypes/i15365.scala [hunk 1]
```diff
@@ -0,0 +1,17 @@
+trait Tagged[U]
```
Comment (dwijnand) [inline tests/neg-custom-args/allow-deep-subtypes/i15365.scala:1]: Btw, you can also just add `// scalac: -Yno-deep-subtypes`

**Comment by odersky** [comment]
OK, I am going to merge this for now, since a more fundamental fix is out of reach.

**Pull request merged**